docs/proposals/20230706-yurtappconfigurationreplacement.md

```diff
@@ -0,0 +1,291 @@
+<img src = "../img/yurtappconfigurationreplacement/Inspiration.png" width="600">
+
+### YurtAppConfigurationReplacement API
+1. YurtAppConfigurationReplacement needs to be bound to YurtAppSet/YurtAppDaemon. 
+Considering that there are multiple Deployment/StatefulSet per nodepool, as shown below, it must be bound to YurtAppSet/YurtAppDaemon for injection. We use subject field to bind it to YurtAppSet/YurtAppDaemon. 
+2. YurtAppConfigurationReplacement is responsible for injection of Replacement. We only need to create a new YurtAppConfigurationReplacement resource for all nodepools under a YurtAppSet/YurtAppDaemon resource. 
+
+```go
+// ImageItem specifies the corresponding container and the claimed image
+type ImageItem struct {
+	// ContainerName represents name of the container 
+	// in which the Image will be replaced
+	ContainerName string `json:"containerName"`
+	// ImageClaim represents the claimed image name 
+	// which is injected into the container above
+	ImageClaim string `json:"imageClaim"`
+}
+
+// EnvItem specifies the corresponding container and the claimed env
+type EnvItem struct {
+	// ContainerName represents name of the container 
+	// in which the env will be replaced
+	ContainerName string `json:"containerName"`
+	// EnvClaim represents the detailed enviroment variables container contains
+	EnvClaim map[string]string `json:"envClaim"`
+}
+
+// PersistentVolumeClaimItem specifies the corresponding container and the claimed pvc
+type PersistentVolumeClaimItem struct {
+	// PVCSource represents pvcClaim name.
+	PVCSource string `json:"pvcSource"`
+	// PVCTarget represents the PVC corresponding to the volume above.
+	// PVCTarget supprot advanced features like wildcard. 
+	// By naming pvc as pvcName-{{nodepool}}, all pvc can be injected at once.
+	PVCTarget string `json:"pvcTarget"`
+}
+
+// ConfigMapItem specifies the corresponding containerName and the claimed configMap
+type ConfigMapItem struct {
+	// ConfigMapSource represents configMap name
+	ConfigMapSource string `json:"configMapSource"`
+	// ConfigMapTarget represents the ConfigMap corresponding to the volume above. 
+	// ConfigMapTarget supprot advanced features like wildcard. 
+	// By naming configMap as configMapName-{{nodepool}}, all configMap can be injected at once.
+	ConfigMapTarget string `json:"configMapTarget"`
+}
+
+type SecretItem struct {
+	// SecretSource represents secret name.
+	SecretSource string `json:"secretSource"`
+	// SecretTarget represents the Secret corresponding to the volume above.
+	// SecretTarget supprot advanced features like wildcard. 
+	// By naming secret as secretName-{{nodepool}}, all secret can be injected at once.
+	SecretTarget string `json:"secretTarget"`
+}
+
+// Item represents configuration to be injected.
+// Only one of its members may be specified.
```

Comment:
What does specifying one of its members mean here? Can I overwrite container images and environment variables at the same time?

Reply:
The reason only one can be specified here is mainly for the sake of a clear hierarchy. Each Item is responsible for only one field. Otherwise, some items contain multiple fields and some have only one field, which is not user-friendly. The relevant validations are performed in the webhook. I think this will be better.

Reply:
Thank you😸. I understand. But I'm a little confused. What should webhook do if I specify multiple items for configmap? What if there are multiple image items?🧐

Reply:
Validation webhook will only do validation for an item containing one element. 
If there are multiple items that all are configMapItem, and their configMapSource are similar, the finally injected configMap will be the one that appears last. 
For image, we can have multiple items to configure different images based on different containerName. If their containerName are similar, the image appears last will work.

Reply:
Thank you. I get it.